cpp/test/mg/pca.cu

```diff
@@ -0,0 +1,214 @@
+                << " matrix" << std::endl;
+    }
+
+    // Prepare X matrix
+    std::vector<Matrix::RankSizePair*> totalPartsToRanks;
+    for (int i = 0; i < params.partSizes.size(); i++) {
+      Matrix::RankSizePair* rspt = new Matrix::RankSizePair(
+        params.ranksOwners[i] % totalRanks, params.partSizes[i]);
+      totalPartsToRanks.push_back(rspt);
+    }
+    Matrix::PartDescriptor desc(params.M, params.N, totalPartsToRanks,
+                                comm.getRank(), params.layout);
+    std::vector<Matrix::Data<T>*> inParts;
+    Matrix::opg::allocate(h, inParts, desc, myRank, stream);
+    Matrix::opg::randomize(h, r, inParts, desc, myRank, stream, T(10.0),
+                           T(20.0));
+    h.waitOnUserStream();
+
+    prmsPCA.n_rows = params.M;
+    prmsPCA.n_cols = params.N;
+    prmsPCA.n_components = params.N_components;
+    prmsPCA.whiten = false;
+    prmsPCA.n_iterations = 100;
+    prmsPCA.tol = 0.01;
+    prmsPCA.algorithm = params.algorithm;
+
+    components = (T*)allocator->allocate(
+      prmsPCA.n_components * prmsPCA.n_cols * sizeof(T), stream);
+
+    explained_var =
+      (T*)allocator->allocate(prmsPCA.n_components * sizeof(T), stream);
+
+    explained_var_ratio =
+      (T*)allocator->allocate(prmsPCA.n_components * sizeof(T), stream);
+
+    singular_vals =
+      (T*)allocator->allocate(prmsPCA.n_components * sizeof(T), stream);
+
+    mu = (T*)allocator->allocate(prmsPCA.n_cols * sizeof(T), stream);
+
+    noise_vars =
+      (T*)allocator->allocate(prmsPCA.n_components * sizeof(T), stream);
+
+    ML::PCA::opg::fit(*handle, inParts, desc, components, explained_var,
+                      explained_var_ratio, singular_vals, mu, noise_vars,
+                      prmsPCA, false);
+
+    Matrix::opg::deallocate(h, inParts, desc, myRank, stream);
+  }
+
+  void TearDown() {
+    const std::shared_ptr<deviceAllocator> allocator =
+      handle->getDeviceAllocator();
+    cudaStream_t stream = handle->getStream();
+
+    allocator->deallocate(
+      components, prmsPCA.n_components * prmsPCA.n_cols * sizeof(T), stream);
+
+    allocator->deallocate(explained_var, prmsPCA.n_components * sizeof(T),
+                          stream);
+
+    allocator->deallocate(explained_var_ratio, prmsPCA.n_components * sizeof(T),
+                          stream);
+
+    allocator->deallocate(singular_vals, prmsPCA.n_components * sizeof(T),
+                          stream);
+
+    allocator->deallocate(mu, prmsPCA.n_cols * sizeof(T), stream);
+
+    allocator->deallocate(noise_vars, prmsPCA.n_components * sizeof(T), stream);
+
+    delete handle;
+  }
+
+ protected:
+  PCAOpgParams params;
+  ML::cumlHandle* handle;
+  cudaStream_t stream;
+  int myRank;
+  int totalRanks;
+  ML::paramsPCAMG prmsPCA;
+  T *components, *explained_var, *explained_var_ratio, *singular_vals, *mu,
+    *noise_vars;
+};
+
+const std::vector<PCAOpgParams> inputs = {{20,
```

Comment:
Wow, did clang-format do this? Seems like it would be easier to read this if it was one record per row, rather than one element.

Reply:
Yes, God, I did not even see this one